pkg/pf/tests/internal/cross-tests/diff.go

```diff
@@ -0,0 +1,175 @@
+			},
+		},
+	}
+}
+
+// Diff will assert that given two resource configurations, the diffs are the same
+// when computed by Terraform and Pulumi.
+//
+// Diff should be safe to run in parallel.
+func Diff(t *testing.T, schema rschema.Schema, tfConfig1, tfConfig2 map[string]cty.Value, options ...DiffOption) crosstestsimpl.DiffResult {
+	skipUnlessLinux(t)
+
+	var opts diffOpts
+	for _, f := range options {
+		f(&opts)
+	}
+
+	// By default, logs only show when they are on a failed test. By logging to
+	// topLevelT, we can log items to be shown if downstream tests fail.
+	topLevelT := t
+
+	prov := pb.NewProvider(pb.NewProviderArgs{
+		AllResources: []pb.Resource{{
+			Name:           "test",
+			ResourceSchema: schema,
+		}},
+	})
+
+	shimProvider := tfbridge.ShimProvider(prov)
+
+	var tfOut string
+	var pulumiOut string
+	var tfChanges tfcheck.TFChange
+	var pulumiRes auto.UpResult
+	var diffResponse crosstestsimpl.PulumiDiffResp
+	t.Run("tf", func(t *testing.T) {
+		defer propagateSkip(topLevelT, t)
+		var hcl1 bytes.Buffer
+
+		err := WritePF(&hcl1).Resource(schema, "testprovider_test", "res", tfConfig1)
+		require.NoError(t, err)
+
+		driver := tfcheck.NewTfDriver(t, t.TempDir(), prov.TypeName, prov)
+
+		driver.Write(t, hcl1.String())
+		plan, err := driver.Plan(t)
+		require.NoError(t, err)
+		err = driver.Apply(t, plan)
+		require.NoError(t, err)
+
+		var hcl2 bytes.Buffer
+		err = WritePF(&hcl2).Resource(schema, "testprovider_test", "res", tfConfig2)
+		require.NoError(t, err)
+		driver.Write(t, hcl2.String())
+		plan, err = driver.Plan(t)
+		require.NoError(t, err)
+		tfChanges = driver.ParseChangesFromTFPlan(plan)
+		tfOut = plan.StdOut
+	})
+
+	t.Run("bridged", func(t *testing.T) {
+		defer propagateSkip(topLevelT, t)
+
+		puConfig1 := crosstestsimpl.InferPulumiValue(t,
+			shimProvider.ResourcesMap().Get("testprovider_test").Schema(),
+			opts.resourceInfo,
+			cty.ObjectVal(tfConfig1),
+		)
+		pulumiYaml1 := yamlResource(t, puConfig1)
+
+		puConfig2 := crosstestsimpl.InferPulumiValue(t,
+			shimProvider.ResourcesMap().Get("testprovider_test").Schema(),
+			opts.resourceInfo,
+			cty.ObjectVal(tfConfig2),
+		)
+		pulumiYaml2 := yamlResource(t, puConfig2)
+
+		bytes, err := yaml.Marshal(pulumiYaml1)
+		require.NoError(t, err)
+		topLevelT.Logf("Pulumi.yaml:\n%s", string(bytes))
+
+		pt, err := pulcheck.PulCheck(t, bridgedProvider(prov), string(bytes))
+		require.NoError(t, err)
+		pt.Up(t)
+
+		bytes, err = yaml.Marshal(pulumiYaml2)
+		require.NoError(t, err)
+		topLevelT.Logf("Pulumi.yaml:\n%s", string(bytes))
+		pt.WritePulumiYaml(t, string(bytes))
+
+		previewRes := pt.Preview(t)
+		pulumiOut = previewRes.StdOut
+
+		pulumiRes = pt.Up(t)
+
+		diffResponse = crosstestsimpl.PulumiDiffResp{}
+		for _, entry := range pt.GrpcLog(t).Entries {
+			if entry.Method == "/pulumirpc.ResourceProvider/Diff" {
+				err := json.Unmarshal(entry.Response, &diffResponse)
+				require.NoError(t, err)
+			}
+		}
```

Comment:
Let's make sure that one and only one diff gets found here.
```suggestion
		var diffResponse struct{ 
			resp crosstestsimpl.PulumiDiffResp{}
			found bool
		}
		for _, entry := range pt.GrpcLog(t).Entries {
			if entry.Method == "/pulumirpc.ResourceProvider/Diff" {
			        require.False(diffResponse.found)
				err := json.Unmarshal(entry.Response, &diffResponse.resp)
				require.NoError(t, err)
				diffResponse.found = true
			}
		}
		require.True(diffResponse.found)
```